cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc

```diff
@@ -1,17 +1,165 @@
+      }
+    }
 
+    array->close();
+    os.flush();
+    filebuf.close();
+  }
+  catch (const std::exception& e) {
+    std::cout << "ArrayUtil::export_file_to_path, caught error:" << e.what() << std::endl;
+    return -1;
+  }
+  catch (...) {
+    std::cout << "ArrayUtil::export_file_to_path, caught unknown error!" << std::endl;
+    return -2;
+  }
 
+  return 0;
+}
  
+int ArrayUtil::save_file_from_path(const std::string& file_uri, const std::string& input_path, const std::string& mime_type, const std::string& mime_encoding, const std::shared_ptr<tiledb::Context>& ctx) {
+  const std::string METADATA_SIZE_KEY = "file_size";
+  const std::string FILE_DIMENSION_NAME = "position";
+  const std::string FILE_ATTRIBUTE_NAME = "contents";
+  const std::string FILE_METADATA_MIME_TYPE_KEY = "mime";
+  const std::string FILE_METADATA_MIME_ENCODING_KEY = "mime_encoding";
+  const std::string METADATA_ORIGINAL_FILE_NAME = "original_file_name";
+  try {
+    if (input_path.empty()) {
+      std::cout << "ArrayUtil::save_file_from_path, input_path can not be empty!" << std::endl;
+      return -1;
+    }
+    std::shared_ptr<tiledb::VFS> vfs = std::shared_ptr<tiledb::VFS>(new tiledb::VFS(ctx));
+    tiledb::VFS::filebuf filebuf = tiledb::VFS::filebuf(vfs);
+    filebuf.open(input_path, std::ios::in);
+    std::istream os(&filebuf);
+    uint64_t nbytes = vfs->file_size(input_path);
+
+
+    uint64_t tile_extent = 1024;
+    if (nbytes > (1024ULL * 1024ULL * 1024ULL * 10ULL)) {
+      tile_extent = 1024ULL * 1024 * 100;
+    }
+    else if (nbytes > (1024UL * 1024 * 100)) {
+      tile_extent = 1024ULL * 1024 * 1;
+    }
+    else if (nbytes > (1024UL * 1024 * 1)) {
+      tile_extent = 1024ULL * 256;
+    }
+
+    std::shared_ptr<tiledb::ArraySchema> file_schema = std::shared_ptr<tiledb::ArraySchema>(new tiledb::ArraySchema(ctx, tiledb::ArrayType::TILEDB_DENSE));
+    tiledb::Domain dom(ctx);
+    std::array<uint64_t, 2> dim_domain = { 0,std::numeric_limits<uint64_t>::max() - tile_extent -2 };
+    tiledb::Dimension dim = tiledb::Dimension::create(ctx, FILE_DIMENSION_NAME, dim_domain, tile_extent);
+    dom.add_dimension(dim);
+    file_schema->set_domain(dom);
+    tiledb::Attribute attr = tiledb::Attribute::create_attribute(ctx, "contents", tiledb::DataType::TILEDB_UINT8);
+    attr.set_cell_val_num(1);
+    file_schema->add_attribute(attr);
+
+    tiledb::Array::create(file_uri, file_schema);
+
+    std::shared_ptr<Array> array = std::shared_ptr<Array>(new Array(ctx, file_uri, tiledb::QueryType::TILEDB_WRITE));
+    Query query(ctx, array, tiledb::QueryType::TILEDB_WRITE);
+
+    std::vector<uint8_t> contents;
+    contents.resize(nbytes);
+    std::array<uint64_t, 2> subarray = { 0,nbytes - 1 };
+    query.set_layout(tiledb::LayoutType::TILEDB_ROW_MAJOR);
+    query.set_subarray(subarray);
+    os.read((char*)contents.data(), nbytes);
+      
+    query.set_buffer(FILE_ATTRIBUTE_NAME, contents);
+    query.submit();
+    query.finalize();
```

Comment:
We need to check the status here, because `finalize` resets the status.

Reply:
Read queries should not call finalize

Reply:
Yes, this is from another function save_file_from_path. I tested with a very simple png image file(Sample.png, about 1Mb). First, use save_file_from_path to save the png file to a tiledb array., then use export_file_to_path to export from the tiledb array to another png file(Sample_exported.png). The size and content of the original file and exported file are the same.

Reply:
We need to check `query_stats` after the write, _before_ calling `finalize`.

Reply:
Bump

Reply:
Are we checking if query_status is INCOMPLETE and do a while loop to continue write?

Reply:
No looping here, this is the write path, but we need to check `query_status == COMPLETE` or tell the user that their write did not complete.

Reply:
added the output when the query is not complete